libres/lib/include/ert/analysis/update.hpp

```diff
@@ -1,14 +1,50 @@
 #include <stdexcept>
-#define HAVE_THREAD_POOL 1
 #include <ert/enkf/enkf_fs.hpp>
 #include <ert/enkf/obs_data.hpp>
 #include <ert/enkf/local_updatestep.hpp>
 #include <ert/enkf/ensemble_config.hpp>
-
-#include <ert/enkf/enkf_state.hpp>
+#include <ert/enkf/analysis_config.hpp>
 #include <ert/enkf/enkf_obs.hpp>
+#include <ert/util/rng.hpp>
 
 namespace analysis {
+/**
+ * Container for all data required for performing an update step.
+ * Data consists of 5 matrices and a list of pairs of rowscaling and matrix.
+ * objects mask describing the observations which
+ * are active. In addition a flag has_observations which is used to determine wheter
+ * it is possible to do an update step.
+*/
+class update_data_type : public std::enable_shared_from_this<update_data_type> {
+public:
+    update_data_type() = default;
+    update_data_type(
+        matrix_type *S_in, matrix_type *E_in, matrix_type *D_in,
+        matrix_type *R_in, std::optional<Eigen::MatrixXd> A_in,
+        std::vector<std::pair<matrix_type *, std::shared_ptr<RowScaling>>>
+            A_with_rowscaling_in,
+        const std::vector<bool> &obs_mask_in)
+        : obs_mask(obs_mask_in) {
+        S = *S_in;
```

Comment:
This should take value arguments and move from - i.e. 
```C++
update_date_type(matrix_type S_in, ....) 
: S(std::move(S_in))
{}
```
and the matrix created at callsite must be freed - as it is now there is a quite significant leak (the move will of course reduce the leak a lot).